private/buf/bufwire/module_config_reader.go

```diff
@@ -205,6 +216,99 @@ func (m *moduleConfigReader) getModuleModuleConfig(
 	return newModuleConfig(module, config, nil /* Workspaces aren't supported for ModuleRefs */), nil
 }
 
+func (m *moduleConfigReader) getProtoFileModuleSourceConfigs(
+	ctx context.Context,
+	container app.EnvStdinContainer,
+	sourceRef buffetch.SourceRef,
+	workspaceBuilder bufwork.WorkspaceBuilder,
+	configOverride string,
+	externalDirOrFilePaths []string,
+	externalDirOrFilePathsAllowNotExist bool,
+) (_ []ModuleConfig, retErr error) {
+	readBucketCloser, err := m.fetchReader.GetSourceBucket(ctx, container, sourceRef)
+	if err != nil {
+		return nil, err
+	}
+	workspaceConfigs := map[string]struct{}{}
+	for _, workspaceConfigFile := range bufwork.AllConfigFilePaths {
+		workspaceConfigs[workspaceConfigFile] = struct{}{}
+	}
+	moduleConfigs := map[string]struct{}{}
+	for _, moduleConfigFile := range bufconfig.AllConfigFilePaths {
+		moduleConfigs[moduleConfigFile] = struct{}{}
+	}
+	terminateFileProvider := readBucketCloser.TerminateFileProvider()
+	workspaceConfigDirectory := ""
+	moduleConfigDirectory := ""
+	for _, terminateFile := range terminateFileProvider.GetTerminateFiles() {
+		if terminateFile != nil {
```

Comment:
Well... we use `make([]TerminateFile, len(terminateFileNames)` when we instantiate the slice for terminate files, so I believe it would range over the slice, even if the terminate file did not end up being populated: https://play.golang.org/p/hV1Czb77_Pg Thus we need to do a `nil` check against the interface.

Reply:
Even if that's the case, we should see how we can refactor things so that we never let this happen. Otherwise users of this abstraction will need to be very careful to not `nil`-dereference the value which will lead to mistakes.

In other words, if I get at least one result from `GetTerminateFiles`, I should be able to access its values without checking if it's `nil`. For example, imagine if I had to check if the `storage.ObjectInfo` was non-`nil` every time I called `readBucket.Walk`?

I realize we aren't initializing it with `make([]TerminateFile, 0, len(terminateFileNames)` since we actually care about indexing into the slice without using `append` (i.e. the position matters here), but there might be something we can do with respect to how its used to initialize the `TerminateFileProvider`. Worst case, we can re-initialize another slice of `[]TerminateFile` without the `nil` elements before it's passed into `newTerminateFileProvider`, but there might be something else we can do to rearrange it a bit.